### ISSUE TITLE
[Materialv2] ThemeResource doesn't work inside a converter

### ISSUE THREAD

**Issue by Lee31416** [unoplatform/Uno.Themes#1157]
### Current behavior

When migrating from material v1 -> v2 every single resource that needs to be changed by a theme now needs to use ThemeResource. However when do so inside a converter it gives an error:

14:26:17:235	Exception thrown: 'Microsoft.UI.Xaml.Markup.XamlParseException' in WinRT.Runtime.dll
14:26:17:235	WinRT information: Failed to assign to property 'x.x.x.NegativeBrush'. [Line: 27 Position: 13]
14:26:17:235	XAML parsing failed.

There's a workaround explained further down

### Expected behavior

The converter should accept ThemeResources.

### How to reproduce it (as minimally and precisely as possible)

1. Download this sample
[UnoSampleConverterBug.zip](https://github.com/unoplatform/uno/files/12295697/UnoSampleConverterBug.zip)
2. Launch it on uwp 
3. Put ThemeResource for one of the 3 brushes inside the converter like this 
![image](https://github.com/unoplatform/uno/assets/72167772/6ca9002f-88aa-4862-9b71-bc8d2fc0a2d3)
4. Launch it again and see it throw an exception.

### Workaround

By having the brush set to static resource in the converter and setting the color of the brush to theme resource we can get the desired effect

### Works on UWP/WinUI

No

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

4.8.15

### Affected platforms

Windows

### IDE

Visual Studio 2019

### IDE version

17.6.2

**Comment by kazo0** [comment]
Yeah you can't use ThemeResource binding for converters :(

**Comment by Lee31416** [comment]
@kazo0 Is that something we can work on or is it set in stone ?

**Comment by jeanplevesque** [comment]
> Yeah you can't use ThemeResource binding for converters :(

@kazo0 What's the solution for that use-case then? Is seems like a software design problem with V2 because that use-case was fine with V1.

**Comment by kazo0** [comment]
> > Yeah you can't use ThemeResource binding for converters :(
> 
> @kazo0 What's the solution for that use-case then? Is seems like a software design problem with V2 because that use-case was fine with V1.

Placing the brushes inside of ThemeDictionaries and using them with ThemeResource is the proper design (https://learn.microsoft.com/en-us/windows/uwp/xaml-platform/themeresource-markup-extension#when-and-how-to-use-themeresource-rather-than-staticresource)

This allows the brushes to change based on application theme as well as allows them to be locally overridden and have the overridden values resolve properly inside of the material styles.

Is there something blocking you from using the StaticResource binding for the converter parameters? Other than the fact that the resource will not be re-evaluated during a theme change?

**Comment by kazo0** [comment]
I may not be understanding the issue but are you telling me you used the ThemeResource binding for converters previously and it worked? I would be surprised as this has never worked on Windows

**Comment by jeanplevesque** [comment]
No, `ThemeResource` references in converters probably never worked.

The regression is that using V1, we can use the PrimaryBrush as a `StaticReference` in a converter and changing theme works. Using V2, that use-case doesn't work because PrimaryBrush is now in a theme dictionary.
On top of that, we can't _migrate_ `StaticResource` in favor of `ThemeResource` because that's just not supported in converters.

Now if you say that `ThemeResource` for brushes is the proper design, that means it's an undocumented breaking change.
The migration documentation (https://platform.uno/docs/articles/external/uno.themes/doc/material-migration.html) should mention this change and also propose an alternative for the converter use-cases.

**Comment by kazo0** [comment]
I'm surprised you didn't run into any issues with the SolidColorBrushes using a ThemeResource binding for the Color as this is not recommended

https://learn.microsoft.com/en-us/windows/apps/design/style/xaml-theme-resources#troubleshooting-theme-resources

![image](https://github.com/unoplatform/Uno.Themes/assets/4793020/a6e96afc-0c56-48be-9b8c-59d92e507380)

**Comment by jeanplevesque** [comment]
Well the brushes were **not** defined in a `ThemeDictionary` right? Only the colors were.
So we would not have that double `ThemeDictionary` thing.

**Comment by kazo0** [comment]
Mmm I see, we should definitely document this change. But I'm not sure what we should do about the converter issue :/ I don't think we should be taking the brushes out of the theme dictionaries

cc @jeromelaban this is just a limitation in Windows that we have no workaround for, right?

**Comment by jeromelaban** [comment]
Here's an update on this topic.

We implemented the Material v2 by putting brushes inside theme dictionaries by following the structure used by Fluent v2. The reason for brushes to be placed in themes to allow for overriding the type of the brush based on theme (e.g. Mica in light/dark, solid in high contrast).

Now with regards to upgrading to v2 and the specific case of converters and the inability to use `ThemeResource` (as mentioned in https://github.com/unoplatform/Uno.Themes/issues/1157#issuecomment-1679480274), this means that you'll need to adjust your converter to be along those lines, which uses strings to do the resolution, instead of static resources:

```csharp
public class MyConverter : IValueConverter
{
    public string? PositiveBrush { get; set; }
    public string? NeutralBrush { get; set; }
    public string? NegativeBrush { get; set; }

    public object? Convert(object value, Type targetType, object parameter, string language)
    {
        return value is double dValue
            ? dValue switch
            {
                > 0 => ResolveResource(PositiveBrush),
                < 0 => ResolveResource(NegativeBrush),
                0 => ResolveResource(NeutralBrush),
                _ => throw new InvalidOperationException()
            }
            : null;
    }

    private object? ResolveResource(string? name)
        => Application.Current.Resources.TryGetValue(name, out object? value) ? value : null;

    public object ConvertBack(object value, Type targetType, object parameter, string language) 
        => throw new NotImplementedException();
}
```

The corresponding XAML:

```xaml
<Page x:Class="testapp02.MainPage"
	xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
	xmlns:local="using:testapp02"
	xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
	xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
	mc:Ignorable="d"
	Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

    <Page.Resources>
        <local:MyConverter x:Key="myConverter"
                           PositiveBrush="SystemControlTransparentBrush"
                           NegativeBrush="SystemControlHighlightListAccentVeryHighBrush"
                           NeutralBrush="SystemControlHighlightListAccentMediumLowBrush" />
    </Page.Resources>

	<StackPanel HorizontalAlignment="Center"
				VerticalAlignment="Center">
        <Slider x:Name="test" Minimum="-10" Maximum="10" />
        <TextBlock AutomationProperties.AutomationId="HelloTextBlock"
					Text="Hello Uno Platform"
                   HorizontalAlignment="Center"
                   Foreground="{x:Bind test.Value, Converter={StaticResource myConverter}, Mode=OneWay}" />
	</StackPanel>
</Page>
```

This will allow for the converter to use theme resources defined from the `app.xaml`, and as the resolution happens on every conversion, theme changes will be applied properly.

We'll be updating the migration documentation to include this technique.

**Comment by jeanplevesque** [comment]
Thanks for the info @jeromelaban! We'll try to go forward with that.
